Replace LazyFlow with FutureFlow (and implement other lazy flows on top of it)

eyalfa opened the pull request:
take a stub at #28016
1. implement futureFlow as a building block.
2. implement lazyFutureFlow and friends in terms of futureFlow and flatMapPrefix
3. fix few failing tests that seem tp have violate the docs

this is still WIP, I think it especially requires a comprehensive test suite for the newly introduce FutureFlow graph stage.

eyalfa committed:
akka-28016__futureFlow: implement FutureFlow graph stage.

eyalfa committed:
akka-28016__futureFlow: introduce FutureFlow as a separate graph stage.

eyalfa committed:
akka-28016__futureFlow: implement some of the Flow factory method using futureFlow and flatMapPrefix

eyalfa committed:
akka-28016__futureFlow: complete FutureFlow with a NeverMat ex when the future fails.

eyalfa committed:
akka-28016__futureFlow: simplify lazyFutureFlow.

eyalfa committed:
akka-28016__futureFlow: fix failing tests. these tests seems to have violated the lazyFutureFlow's docs by expecting the wrong exception.

eyalfa committed:
Merge branch 'master' into akka-28016__futureFlow

# Conflicts:
#	akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

eyalfa commented:
@johanandren , @jrudolph , I'd appreciate if you guys could take a look.

eyalfa committed:
akka-28016__futureFlow: remove deprecated EC.

eyalfa committed:
akka-28016__futureFlow: fmt

eyalfa commented:
> @johanandren , @jrudolph , I'd appreciate if you guys could take a look.

guys?

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,169 @@
+              if (!isClosed(in) && !hasBeenPulled(in)) {
+                pull(in)
+              }
+            }
```
```suggestion
            override def onPull(): Unit = if (!isClosed(in)) tryPull(in)
```

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,169 @@
+            }
+          }
+        }
+        pr.complete {
```
Turn this around and do something like

```scala
Try(Source.fromGraph(subSource.source).viaMat(fl)(Keep.right).to(subSink.sink).run()(subFusingMaterializer)) match {
  case Success(mat) =>
    pr.complete(mat) 
    ...
```

instead of completing and then matching.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,169 @@
+        }
+        pr.future.value.get match {
+          case Success(_) =>
+            if (isAvailable(in)) {
```
This shouldn't happen any more if you don't propagate pulls.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,169 @@
+            if (isAvailable(in)) {
+              subSource.push(grab(in))
+            }
+            initializing.upstreamFailure match {
```
:+1:

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,169 @@
+              case OptionVal.Some(cause) =>
+                subSink.cancel(cause)
+              case OptionVal.None =>
+                if (hasBeenPulled(in)) {
```
It should be `if (isAvailable(out))`, i.e. it has been pulled from out before materialization.

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -665,7 +674,23 @@ object Flow {
+          implicit val ec = akka.dispatch.ExecutionContexts.parasitic
+          val f: Flow[I, O, Future[M]] =
+            futureFlow(create().map(Flow[I].prepend(Source.single(a)).viaMat(_)(Keep.right)))
+          f
```
```suggestion
          futureFlow(create().map(Flow[I].prepend(Source.single(a)).viaMat(_)(Keep.right))(ExecutionContexts.parasitic))
```

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -665,7 +674,23 @@ object Flow {
+          f
+        case Seq() =>
+          val f: Flow[I, O, Future[M]] = Flow[I]
+            .asInstanceOf[Flow[I, O, NotUsed]]
```
Why is that needed?

eyalfa replied:
could also be done by:
`Flow[I].map[O](_ => ???)`
which one do you prefer?

seglo replied:
> Why is that needed?

`O` can't be inferred otherwise.

jrudolph replied:
Yes, agreed.

jrudolph reviewed (commented):
Looks reasonable. I guess the assumption is that the existing LazyFlowSpec should cover all cases. I didn't review whether these are reasonable or complete enough. The one semantic thing I find questionable is whether the stage should eagerly pull before the flow has been materialized.

Some cosmetic comments:

 * some detailed comments/suggestions below
 * remove commented code, extra braces and printlns everywhere

jrudolph commented:
> > @johanandren , @jrudolph , I'd appreciate if you guys could take a look.
> 
> guys?

Sorry for the delay, @eyalfa. We are somewhat busy right now and bandwidth-limited so we might not get to PRs as quickly as we would like.

eyalfa commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazyFlowSpec.scala:
```diff
@@ -126,7 +126,8 @@ class LazyFlowSpec extends StreamSpec("""
       val deferredMatVal = result._1
```
no, I definitely intend to add a FutureFlow specific suite.

eyalfa commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,169 @@
+
+      object initializing extends InHandler with OutHandler {
+        override def onPush(): Unit = {
+          //just leave it in the port
```
good point, if I remember the pull instead of actually forwarding it I wouldn't have to buffer elements or rely on the port to keep it for me.
b.t.w my impl assumes the element is kept in the port as long as it's not pulled again, what happens in case of an upstream error? which btw may happen out of band... I guess this will remove the element from the port, right?
This also relates to some conversation I had with @johanandren in [this comment](https://github.com/akka/akka/issues/28016#issuecomment-595059713) on #28204 
so if we never pull upstream before the future flow was materialized we don't have to mess with this, right? if an out of band failure happens we won't have any pending upstream pull hence we definitely won't have a pending element in the port.

how do you suggest `flatMapPrefix` to handle this? we must pull upstream there, and we explicitly buffer pushed elements, but currently in case of an 'early' upstream failure we don't materialize the flow. this theoretically prevents users from providing flows with 'recover' operators (though they will handle 'late' failures)

jrudolph replied:
> so if we never pull upstream before the future flow was materialized we don't have to mess with this, right? if an out of band failure happens we won't have any pending upstream pull hence we definitely won't have a pending element in the port.

Yes, failures will clear the port which usually is fine since you the failure should be propagated as quick as possible.

jrudolph replied:
> how do you suggest `flatMapPrefix` to handle this? we must pull upstream there, and we explicitly buffer pushed elements, but currently in case of an 'early' upstream failure we don't materialize the flow. this theoretically prevents users from providing flows with 'recover' operators (though they will handle 'late' failures)

I see, that's a bit of a weird edge case. If we wanted to give `flatMapPrefix` users more control we would then have to think about how to represent the case where only part of the prefix has been collected and then the stream failed. Not sure we want to go into that right now? Maybe we wait until someone asks for that feature.

eyalfa replied:
I think a custom attribute is the way to go, but probably not now 😎

eyalfa commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,169 @@
+          upstreamFailure = OptionVal.Some(ex)
+        }
+
+        override def onPull(): Unit = pull(in)
```
I tend to agree, but not sure what u mean by 'decision'. I think we should record the pull and once the future flow materializes we should signal it with the pull.

jrudolph replied:
What I meant is that when the flow is materialized, yes, we should signal it with the pull and then the flow can decide whether to propagate it to its other end or just produce elements itself.

eyalfa commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,169 @@
+            failStage(exception)
+          case Success(flow) =>
+            //materialize flow, connect to inlets, feed with potential events and set handlers
+            val initialized = new Initialized(flow)
```
not sure I'm following here, can u please elaborate?

seglo replied:
I think @jrudolph may be suggesting moving the logic of `onFuture` (and calling it `connect`?) back up to enclosing `GraphStageLogic` since you're interacting with its members and not those of the `initializing` type.

eyalfa commented:
@jrudolph , thanks for your review! definitely made few important points!
can you please suggest some tests u'd like to see for the new `FutureFlow` stage?
I can think of few out of the top of my head:
1. happy case when Future is already completed (Future.successful)
2. happy case for late materialization
3. failed future (Future.failed)
4. late failed future
5. upstream failure when future is completed
6. upstream failure when future is not yet complete
7. 5 and 6 with a flow that recovers from failures
8. flow that 'ignores upstream completion (concat/sinkAndSource)
9. flow that ignores downstream completion (i.e. broadcast)

I'd appreciate any input from your side.

eyalfa commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,169 @@
+          pr.failure(new NeverMaterializedException())
+        }
+        super.postStop()
+      }
```
when working on flatMapPrefix, @johanandren  commented that the preferred from was testing `isComplete` first in order to avoid the cost of allocating an exception and capturing the current stack trace (NeverMaterializedException does not extend `NoStackTrace`)

eyalfa committed:
akka-28016__futureFlow: address most of @jrudolph's comments.

eyalfa commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,169 @@
+      setHandlers(in, out, initializing)
+
+      override def preStart(): Unit = {
+        super.preStart()
```
explicit is better than implicit 😄

jrudolph replied:
It's just not needed ;)

seglo replied:
Seems redundant. `super.preStart` does nothing.

eyalfa committed:
akka-28016__futureFlow: one more comment addressed.

eyalfa committed:
akka-28016__futureFlow: very initial suite for futureFlow

eyalfa committed:
akka-28016__futureFlow: fmt

eyalfa committed:
akka-28016__futureFlow: few more tests

eyalfa committed:
akka-28016__futureFlow: some more tests, one actual fix.

eyalfa committed:
akka-28016__futureFlow: fmt

eyalfa committed:
akka-28016__futureFlow: add tests for materialization failures

eyalfa committed:
akka-28016__futureFlow: add tests for flow failures.

eyalfa committed:
akka-28016__futureFlow: add tests for downstream completion handling and abrupt termination.

eyalfa committed:
akka-28016__futureFlow: fix a failing test.

eyalfa commented:
CC: @jrudolph , @seglo

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFutureFlowSpec.scala:
```diff
@@ -0,0 +1,411 @@
+        .toMat(Sink.seq)(Keep.both)
+        .run()
+
+      fSeq1.futureValue should be(empty)
```
Is this true deterministically? The question is if `alsoTo` might already send elements to the second branch of the stream before it receives the cancellation. But maybe it just works out because there are no hidden buffers. In any case, it seems a bit brittle to expect that so there should be a at least a comment that warns about it.

eyalfa replied:
let's make sure I understand your concerns, you're talking about the fact there's nop guaranteed order between upstream events and downstream cancellation/failure?
I think it doesn't matter in this case since takeWhile/take will fire a cancellation only after requesting and receiving n messages.

I think your concerns hold for downstream with async boundaries/behaviour or buffering where a cancellation may be fired out of band, but not for this specific test.

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFutureFlowSpec.scala:
```diff
@@ -0,0 +1,411 @@
+        .viaMat {
+          Flow.futureFlow {
+            Future.successful {
+              Flow[Int].orElse(Source.single(99))
```
Didn't know about `orElse`, good example here :)

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFutureFlowSpec.scala:
```diff
@@ -0,0 +1,411 @@
+      fSeq.failed.futureValue should equal(TE("fail on 5"))
+    }
+
+    "allow flow to handle downstream completion with a completed future" in assertAllStagesStopped {
```
Is there a reason that there's no "late future" variant for this one?

eyalfa replied:
right, I'll add it.

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFutureFlowSpec.scala:
```diff
@@ -0,0 +1,411 @@
+
+import scala.concurrent.{ Future, Promise }
+
+class FlowFutureFlowSpec extends StreamSpec {
```
Looks pretty comprehensive :+1:

jrudolph reviewed (commented):
A few more comments

eyalfa committed:
akka-28016__futureFlow: add missing test

seglo commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+
+@InternalApi private[akka] final class FutureFlow[In, Out, M](future: Future[Flow[In, Out, M]])
```
```suggestion
@InternalApi private[akka] final class FutureFlow[In, Out, M](futureFlow: Future[Flow[In, Out, M]])
```

seglo commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+      override def preStart(): Unit = {
+        super.preStart()
+        future.value match {
+          case Some(tr) =>
```
```suggestion
          case Some(tryFlow) =>
```

seglo commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+
+      override def preStart(): Unit = {
+        super.preStart()
+        future.value match {
```
I like the comment from `LazyFlow` on why this is required, but maybe that's just because I'm not as familiar with the trick.

```
          // check if the stage is still in need for the lazy flow
          // (there could have been an onUpstreamFailure or onDownstreamFinish in the meantime that has completed the promise)
```

seglo commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+          case Success(flow) =>
+            //materialize flow, connect to inlets, feed with potential events and set handlers
+            val initialized = new Initialized(flow)
+            setHandlers(in, out, initialized)
```
Neat pattern :+1:

seglo commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+        super.postStop()
+      }
+
+      object initializing extends InHandler with OutHandler {
```
```suggestion
      object Initializing extends InHandler with OutHandler {
```

eyalfa replied:
I think of this as a 'shortcut' for a class + val definition, so I'd prefer leaving it with the val naming notation.

eyalfa replied:
@seglo do u insist on this?

seglo replied:
Yes, I think we should make the change. I've never seen lowerCamelCase for object name's, other than for package objects. I'll admit I don't know if this formatting is violated elsewhere in the akka codebase, but the scala-lang.org Scala style guide says:

> Object names are like class names (upper camel case).
>
> An exception is when mimicking a package or function. This isn’t common.

https://docs.scala-lang.org/style/naming-conventions.html

seglo commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+            setKeepGoing(false)
+        }
+      }
+      class Initialized(fl: Flow[In, Out, M]) extends InHandler with OutHandler {
```
```suggestion
      class Initialized(flow: Flow[In, Out, M]) extends InHandler with OutHandler {
```

seglo commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        override def onUpstreamFinish(): Unit = {
+          //will later be propagated to the materialized flow
```
Shouldn't this be handled during initialization too?

eyalfa replied:
it is, it just doesn't require any specific state on top of `isClosed(in)`

seglo replied:
I'm not sure I understand.  How do we close the stage when its in `initializing` when the upstream completes?

eyalfa replied:
when initializing stage set's keepGoing=true, then when the future completes the stage changes state, materializes the flow, connects its ports and forwards the signals.
specifically it checks isClosed(in) and closes the subSource in such cases.

seglo replied:
I'm not the signals would be forwarded after new handlers are set.  `onUpstreamFinish` will only be called once, so if that happens before the future completes then we'll miss the signal.

eyalfa replied:
when onUpstreamFinish is called before future completion we do not close the stage, furthermore even if downstream is completed we do **not** close the stage (due to `setKeepGoing(true)`).
at some point the future completes and the flow is materialized. at this point the 'stashed' signals are forwarded to the materialized flow, this may include an upstream completion or failure, and downstream pull cancellation or failure.

another related point, it is possible for the 'outer' stage to complete while the 'inner' flow keeps running. this is achieved by `setKeepGoinf(false)` after materializing the nested flow and forwarding the signals, if both upstream and downstream are closed at this point the outer stage will complete but it's up to the nested flow to decide how to handle the closed ports. you can see examples in the tests.

seglo replied:
> at some point the future completes and the flow is materialized. at this point the 'stashed' signals are forwarded to the materialized flow, this may include an upstream completion or failure, and downstream pull cancellation or failure.

I'm unfamiliar with this behaviour, but I'll dig into it for my own understanding.  I'll follow up.

jrudolph replied:
@seglo, see line 120 - 126. The stashing is part manual / part built-in. If there's a failure it's explicitly saved in `upstreamFailure`. If it has been completed, `isClosed(in)` will return true. So, no magic involved.

eyalfa replied:
thanks @jrudolph, @seglo notice this code is now migrated into the connect method you guys requested.

seglo replied:
Thanks for the clarification.  I see it now.  Thanks.

seglo reviewed (commented):
I like this direction. Let's drop all the commented/dead code and get more eyes on it. I left mostly cosmetic naming suggestions. I haven't dug into the testsuite in detail yet.

eyalfa committed:
akka-28016__futureFlow: address review comments.

eyalfa committed:
akka-28016__futureFlow: compilation fixes + scalafmtAll.

eyalfa commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -584,7 +584,12 @@ object Flow {
+    /*Flow.fromGraph(new LazyFlow[I, O, M](flowFactory)).mapMaterializedValue(_ => fallback())*/ Flow[I]
+      .flatMapPrefix(1) {
+        case Seq(a) => futureFlow(flowFactory(a)).mapMaterializedValue(_ => NotUsed)
+        case Seq()  => Flow[I].asInstanceOf[Flow[I, O, NotUsed]]
```
`flatMapPrefix` will feed the given function with a `Seq`, not guaranteed to be a `List`

seglo replied:
```
Welcome to Scala 2.13.1 (OpenJDK 64-Bit Server VM, Java 1.8.0_242).
Type in expressions for evaluation. Or try :help.

scala> scala.collection.immutable.Seq() == Nil
res0: Boolean = true
```

eyalfa replied:
does this hold for scala 2.12 as well?
I think it's a bit odd to treat one case as `Seq(a)` and the other as `Nil`.

i'll change this if u insist, lmk.

seglo replied:
Yes it does.  I think it's more idiomatic to match on `Nil`.

eyalfa commented:
@jrudolph , @seglo ,
Not sure if this PR is the right place for discussing this but; I've started using flatMapPrefix in my production code and found a surprising side effect: downstream cancellation is delayed, this is quite surprising when this cancellation is actually a failure.
The reasons I bring this up:
1. same concern applies to `FutureFlow` as well.
2. The lazy variants modified in this PR are built on top `flatMapPrefix` and `FutureFlow`, therefore they are affected as well.

I think we should introduce some reasonable default behaviour (i.e. propagate downstream failures but not 'regular' cancellations) and enable configuration by attributes.

WDUS?

eyalfa commented:
guys, seems like a mima issue: https://travis-ci.org/github/akka/akka/jobs/668008396#L308-L309
do I have to resurrect the LazyFlow class? is their a way to exclude it from the test? I've checked it was annotated with InternalAPI and it was package private, doesn't it mean it can be safely removed?

seglo commented:
> guys, seems like a mima issue: [travis-ci.org/github/akka/akka/jobs/668008396#L308-L309](https://travis-ci.org/github/akka/akka/jobs/668008396#L308-L309)
> do I have to resurrect the LazyFlow class? is their a way to exclude it from the test? I've checked it was annotated with InternalAPI and it was package private, doesn't it mean it can be safely removed?

Since it's an internal API you can add a mima exception.  Add a new filter file `akka-stream/src/main/mima-filters/2.6.4.backwards.excludes` and add the exception rule from the build:

```
ProblemFilters.exclude[MissingClassProblem]("akka.stream.impl.fusing.LazyFlow")
```

seglo commented:
> Not sure if this PR is the right place for discussing this but; I've started using flatMapPrefix in my production code and found a surprising side effect: downstream cancellation is delayed, this is quite surprising when this cancellation is actually a failure.

I took a peek at `FlatMapPrefix`.  This is just a guess, but since materialize the flow late and forward cancellations through the subSink, and then to the stage, some ticks could happen before the stage is actually cancelled.  Maybe the stage could handle cancelling itself more directly instead of waiting for the propagation through the subSink.

eyalfa commented:
> > Not sure if this PR is the right place for discussing this but; I've started using flatMapPrefix in my production code and found a surprising side effect: downstream cancellation is delayed, this is quite surprising when this cancellation is actually a failure.
> 
> I took a peek at `FlatMapPrefix`. This is just a guess, but since we materialize the flow late and forward cancellations through the subSink, and then to the stage, some ticks could happen before the stage is actually cancelled. Maybe the stage could handle cancelling itself more directly instead of waiting for the propagation through the subSink.

@seglo , if u look at flatMapPrefix test suite u'd see this is actually a feature as it allows 'plugging in' flows with custom handling for upstream/downstream cancellation/failure/completion, thing is there seem to me like there are some common use-cases where this kind of flexibility isn't required and may even be undesired. I think this behaviour should eventually be configurable via attributes, I also tend to believe default behaviour should be failing the 'outer' stage without waiting for the internal stage to be materialized.
The reason /i've mentioned it here was because this stage's design followed the convention formed when working on flatMapPrefix (I think it was actually suggested by @jrudolph) hence it'd surprise users in the same way.

ennru commented:
@eyalfa Thank you for your dedicated work in this!

I took the liberty to give this PR a more describing title.

As a next step, I believe it would be beneficial to expand the operator docs for the touched operators. If we can describe the peculiarities of lazy flows and inits in words and examples we're on a good path.

eyalfa committed:
akka-28016__futureFlow: add mima exception for Lazyflow.

eyalfa committed:
Merge remote-tracking branch 'origin/master' into akka-28016__futureFlow

eyalfa commented:
@seglo , I've tried fixing the MIMA issues I had nut now it fails on others unrelated MIMA issues.
I suspect those were introduced by #28739 .

CC: @iorskovic (author of #28739), @johanandren (reviewer)

ioreskovic commented:
> @seglo , I've tried fixing the MIMA issues I had nut now it fails on others unrelated MIMA issues.
> I suspect those were introduced by #28739 .
> 
> CC: @ioreskovic (author of #28739), @johanandren (reviewer)

I also had MiMa issues, and [these](akka-actor-typed/src/main/mima-filters/2.6.3.backwards.excludes/pr-28739-expose-stashbuffer-capacity.excludes) were the suggested lines to add.

Should I have done something else/more?

eyalfa commented:
> > @seglo , I've tried fixing the MIMA issues I had nut now it fails on others unrelated MIMA issues.
> > I suspect those were introduced by #28739 .
> > CC: @ioreskovic (author of #28739), @johanandren (reviewer)
> 
> I also had MiMa issues, and [these](akka-actor-typed/src/main/mima-filters/2.6.3.backwards.excludes/pr-28739-expose-stashbuffer-capacity.excludes) were the suggested lines to add.
> 
> Should I have done something else/more?

@ioreskovic , I'm not sure. your PR passed CI, mine doesn't. at first it fail on mima issues directly related to my stuff but after adding the lines suggested by @seglo it started failing on the methods you added in your PR, merging master into my branch did not resolve the issue.

seglo commented:
> @ioreskovic , I'm not sure. your PR passed CI, mine doesn't. at first it fail on mima issues directly related to my stuff but after adding the lines suggested by @seglo it started failing on the methods you added in your PR, merging master into my branch did not resolve the issue.

@eyalfa Something went wrong in your merge of master.  In your branch the excludes filter for #28739 `pr-28739-expose-stashbuffer-capacity.excludes` landed in `akka-actor-typed/src/main/mima-filters/2.6.3.backwards.excludes/` when it should be in `akka-actor-typed/src/main/mima-filters/2.6.4.backwards.excludes/` (since it was merged for the upcoming 2.6.5.  When I merged master locally the mima check passed.

Also, I resolved some of the PR review comments. There are still some outstanding I think need your follow up.

ioreskovic commented:
> > @ioreskovic , I'm not sure. your PR passed CI, mine doesn't. at first it fail on mima issues directly related to my stuff but after adding the lines suggested by @seglo it started failing on the methods you added in your PR, merging master into my branch did not resolve the issue.
> 
> @eyalfa Something went wrong in your merge of master. In your branch the excludes filter for #28739 `pr-28739-expose-stashbuffer-capacity.excludes` landed in `akka-actor-typed/src/main/mima-filters/2.6.3.backwards.excludes/` when it should be in `akka-actor-typed/src/main/mima-filters/2.6.4.backwards.excludes/` (since it was merged for the upcoming 2.6.5 release). When I merged master locally the mima check passed.
> 
> Also, I resolved some of the PR review comments. There are still some outstanding I think need your follow up.

Could it be that I placed mima filters in the wrong directory?

seglo commented:
> Could it be that I placed mima filters in the wrong directory?

Possibly, but it's in the correct directory in master right now.

eyalfa commented:
ok, something'd odd with this mima exclusion added by @ioreskovic:

> git show 3ccaee2c18 --name-only  | cat                                                    ✔  7930  10:06:51
> commit 3ccaee2c18ac0c377b028dc46ebdfdf277c0bc84
> Author: Ivan Oreskovic <ivan.oreskovic@gmail.com>
> Date:   Wed Apr 1 14:25:34 2020 +0200
> 
>     Expose capacity property on StashBuffer (#28739)
> 
> akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StashBufferSpec.scala
> akka-actor-typed/src/main/mima-filters/2.6.3.backwards.excludes/pr-28739-expose-stashbuffer-capacity.excludes
> akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/StashBuffer.scala
> akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/StashBuffer.scala

as you can see, it was added in the `2.6.3.backwards.excludes` folder, so my merge went well 😎 

the file added by @ioreskovic also exists in master (as expected):

> git ls-tree master:akka-actor-typed/src/main/mima-filters/2.6.3.backwards.excludes        ✔  7934  10:07:44
> 100644 blob 18cda0248c9a9cb2beae5ea14bf966b64015faf6	28630-ActorRef-ignore.excludes
> 100644 blob 9fe23b3edc2ae25ca9a4714d287ab998e01488a5	pr-28739-expose-stashbuffer-capacity.excludes

seems like there's another commit (34af198bd380de5e1b574fe6e0958cda2398c23d) by @patriknw that added some more exclusions on the same class:

> git show master:akka-actor-typed/src/main/mima-filters/2.6.4.backwards.excludes/issue-28786-stash-clear.excludes | cat
> \# \#28786 add clear to StashBuffer, ok because of DoNotInherit
> ProblemFilters.exclude\[ReversedMissingMethodProblem]("akka.actor.typed.javadsl.StashBuffer.clear")
> ProblemFilters.exclude\[ReversedMissingMethodProblem]("akka.actor.typed.scaladsl.StashBuffer.clear")

could this somehow have broken the build?

@seglo, can you please advice on the right action here?

seglo commented:
I think the last time you merged master the excludes file from #28786 was in the wrong directory, but it has since been moved.  If you fetch the latest and merge master you should be able to run `mimaReportBinaryIssues`.  It worked for me locally.

eyalfa committed:
Merge branch 'master' into akka-28016__futureFlow

eyalfa commented:
> mimaReportBinaryIssues

thanks @seglo , a second pull did the trick.
I'll go over the rest of the comments tomorrow (hopefully)

eyalfa commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,168 @@
+          case Failure(ex) =>
+            innerMatValue.failure(new NeverMaterializedException(ex))
+            failStage(ex)
+        }
```
@seglo , @jrudolph ,
so u basically suggest extracting this section to a `connect()` method?
another option is to create the nested ports (subSource, subSink) in the connect method and pass them to the `Initialized`constructor.

is this the direction you guys suggest?

seglo replied:
I thought @jrudolph was referring to the `Initializing.onFuture` method.  Since this method only references outer stage logic members it should live in the outer stage.  However, I'm not 100% that's what @jrudolph so it's best he respond here.

Is this comment related to https://github.com/akka/akka/pull/28729#discussion_r395285748 ?

jrudolph replied:
Yes, everything that's done in `case Success(flow)` in line 122, should be put into a single `def connect` method. There's no real need to create an `Initialized` state then, the handlers could just be anonymous. If you want to give it a name, it could still be a class inside of the `connect` method that closes over the `subSource`, `subSink`. If you want to keep it really explicit, you can also keep the class at top level and pass those in the constructor. Just make sure for clarity, that no initialization code is run in that state class.

eyalfa committed:
akka-28016__futureFlow: address comment by @seglo.

eyalfa committed:
akka-28016__futureFlow: address comments by @jrudolph and @seglo

eyalfa committed:
akka-28016__futureFlow: scalafmt

eyalfa committed:
Merge branch 'master' into akka-28016__futureFlow

eyalfa committed:
akka-28016__futureFlow: address review comment by @seglo

seglo commented:
Thanks @eyalfa.  I think this is getting close.  I'll do another review tomorrow.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
```diff
@@ -8,21 +8,20 @@ import java.util.concurrent.TimeUnit.NANOSECONDS
 import scala.collection.immutable
 import scala.collection.immutable.VectorBuilder
-import scala.concurrent.{ Future, Promise }
+import scala.concurrent.{ Future }
```
```suggestion
import scala.concurrent.Future
```

jrudolph commented on akka-stream/src/main/mima-filters/2.6.4.backwards.excludes/28729-future-flow.backwards.excludes:
```diff
@@ -0,0 +1,2 @@
+ProblemFilters.exclude[MissingClassProblem]("akka.stream.impl.fusing.LazyFlow")
```
```suggestion
# Changes to internals
ProblemFilters.exclude[MissingClassProblem]("akka.stream.impl.fusing.LazyFlow")
```

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+                if (isClosed(in))
+                  subSource.complete()
+            }
+            Initializing.downstreamCause match {
```
Better pass these into the def for more clarity.

eyalfa replied:
@jrudolph , can u please explain? do u mean explicitly passing the vars as method parameters to the connect method?
perhaps it makes more sense to make the connect method a member of Initializing?

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              subSink.cancel(cause)
+              //super.onDownstreamFinish(cause)
```
Some comments that can be removed also further down.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+          new OutHandler {
+            override def onPull(): Unit = if (!isClosed(in)) tryPull(in)
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
```
Make these handler methods all one-liners without braces where possible. Also further down.

seglo replied:
There's a few more event handlers in `Initializers` that can be made one-liners too.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+              case OptionVal.None =>
+                //todo: should this be invoked before and independently of checking downstreamCause?
+                // in most case if downstream pulls and then closes, the pull is 'lost'. is it possible for some flows to actually care about this? (non-eager broadcast?)
+                if (Initializing.hasBeenPulled) {
```
I don't think we need this flag. You can use `isAvailable(out)` instead.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+                subSink.cancel(cause)
+              case OptionVal.None =>
+                //todo: should this be invoked before and independently of checking downstreamCause?
+                // in most case if downstream pulls and then closes, the pull is 'lost'. is it possible for some flows to actually care about this? (non-eager broadcast?)
```
It's fine like this. It might change behavior in upstream stages but that's expected. Cancellation can happen at any time.

jrudolph reviewed (commented):
:+1: Afaics only cosmetic things left.

eyalfa committed:
Update akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

Co-Authored-By: Johannes Rudolph <johannes.rudolph@gmail.com>

eyalfa committed:
Update akka-stream/src/main/mima-filters/2.6.4.backwards.excludes/28729-future-flow.backwards.excludes

Co-Authored-By: Johannes Rudolph <johannes.rudolph@gmail.com>

seglo reviewed (commented):
This is looking really good @eyalfa, great work!  I think the implementation is ready.  A few more things to address to wrap things up.

The PR that inspired this one (#28497) contains a test that verifies `LazyFlow` works as it did in Akka 2.5.  I ran the test in this PR successfully, can you review it and see if it's worth adding to this PR if it's not superseded by your other test coverage?  Once we're certain we have this coverage I think it would be safe to close #28497 in favour of this PR.  The applicable test:
	
https://github.com/aperkins1310/akka/blob/lazy-flow/akka-stream-tests/src/test/scala/akka/stream/scaladsl/LazyFlowSpec.scala#L161..L179

I hesitate to add more scope to this PR, but I think it would be appropriate to provide more documentation around the future and lazy flow operator variants.  In #28707 @johanandren made a statement about providing more documentation to explain them better.

> It would also be good if we ended up with a overview documentation page for these operators.

Would you be open to adding content to a new section in the [Akka Streams documentation](https://doc.akka.io/docs/akka/current/stream/index.html)?  

I think that it would also be great if you could review the [operator docs](https://doc.akka.io/docs/akka/current/stream/operators/index.html) for the affected operators, but I think this would be fine to do as a follow up PR or by someone else.

eyalfa committed:
akka-28016__futureFlow: address comments by @jrudolph

eyalfa committed:
akka-28016__futureFlow: add testcase requested by @seglo

eyalfa committed:
akka-28016__futureFlow: scalafmt

eyalfa commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -665,7 +670,20 @@ object Flow {
    * '''Cancels when''' downstream cancels
```
@seglo , notice this is one aspect of the operator's behaviour that has slightly changed.
an 'early' cancellation is stashed until the inner flow is materialized, it is then handed over to the materialized flow which can decide on handling the signal.
as mentioned in an earlier comment of mine, this behaviour is inspired by `flatMapPrefix` (and review comments+directions by @jrudolph ), I think we should also provide some fine grained control over this behaviour by specifying attributes. 

given that most flows actually cancels when downstream does, this change of behaviour is usually unnoticeable as long as upstream emits/completes and the future actually completes in a reasonable time.

seglo replied:
Thanks for pointing this out.  I think it would make sense to preserve this behaviour for `lazyFutureFlow` to avoid more possible regressions for existing users.  

Are you suggesting creating a new attribute to control whether operators that use `FutureFlow` should eagerly cancel the stage `onDownstreamFinish(Throwable)` (default case), or pass the signal through the inner stream materialization?  There's already a `CancellationStrategy` attribute, perhaps we could add a strategy to it.

eyalfa replied:
I'm basically talking about an attribute that controls this behaviour, it's needed both for `FutureFlow` and `FlatMapPrefix` as both currently 'intercepts' the downstream completion and delays its handling until there's a nested flow to delegate to. I also think the current behaviour should **not** be the default.
re. the existing `CancellationStrategy` attribute, it seems to be related to the way stages are cancelled so I'm not sure it makes sense to reuse it.
I'm also not sure if the same attribute can be used to control both stages, since it may make it difficult to specify for flows that combine both, on the other hand it seems that these kind of combinations (i.e. the lazyXXX family) would usually apply the same policy to both.

do u think we should implement this as part of this PR or in a separate one? I guess it boils down to the importance of maintaining the old behaviour.
please advice.

jrudolph replied:
I tend to say that relying on the previous behavior is somewhat brittle and already required careful design to be free of race conditions. To observe the new behavior you must have control over 1. when the future is completed and 2. the cancellation is received. In most cases, both of these are asynchronous and one can happen before the other.

The most likely case that I can think of is that you know 1. that cancellation was triggered and will arrive eventually and 2. the future flow is never completed. In particular, there's the case that the future is left hanging uncompleted accidentally. In this case the previous behavior would have prevented leaking the upstream. That sounds significant but even that is somewhat expected as never completing futures is known to lead to leaks.

Any other case how you would observe it?

eyalfa replied:
I think the current design eliminates the race condition, at least as a corruption. the stage is notified by 'wiring' Future.onComplete to an asyncCallback and internal state is updated only in the stage's context 
I guess u can call it a race condition since the stage may become non-deterministic, is this what u mean @jrudolph ? if so do u prefer the current implementation as proposed in this PR?

I think it's also worth noting that most flows simply propagate the downstream cancellation upstream, which results with the in port being closed. one downside of my current implementation is that in this common case the downstream/cancellation/failure signal is 'delayed'. since I believe this is the common case I think it should also be the default behaviour or at least possibly enabled using an attribute.

jrudolph replied:
> I think it's also worth noting that most flows simply propagate the downstream cancellation upstream, which results with the in port being closed. one downside of my current implementation is that in this common case the downstream/cancellation/failure signal is 'delayed'.

That's a good point. Another argument for the old behavior by default would be that you usually don't need the extra flexibility of having the flow react specially on cancellation. In almost all cases, the flow will just propagate the cancellation.

For `flatMapPrefix` it might be a bit different because there you'd might want to execute a side-effect when the prefix is collected regardless of whether the downstream was already cancelled or not.

I talked with @seglo about that whole issue and as much as we'd like to avoid the complexity of introducing an extra attribute/parameter to configure this, it seems it actually is needed.

Because most usages will not depend on holding back cancellation we'd say that the old behavior would be the default (for both `futureFlow` and `flatMapPrefix` for consistency) and if you want the extra flexibility of having the flow handle cancellation you would have to add an attribute as you suggested. It seems the extra complexity in the the stage implementations would be low, mostly a small change where instead of stashing the downstream cause you would eagerly cancel the stage.

WDYT, @eyalfa?

So, the changes needed are:

 * add the new attribute
 * implement the feature in FutureFlow and FlatMapPrefix
 * add/adapt tests to test both behaviors
 * document the attribute in scaladoc
 * (maybe later) make sure the stream operation documentation pages are adapted accordingly (we are thinking about having an extra page to document the features and differences between all those flow construction helpers)

eyalfa replied:
@jrudolph ,
this basically makes sense, few questions come to mind (on a first read, there may be more later on):
1. same attribute for both stages?
2. default should be the 'old behaviour' hence, if downstream cancels before future/prefix completion the stage cancels, otherwise the inner flow takes charge. right?
3. another possibility is 'going all the way' by always closing upstream upon downstream cancellation, even if the nested flow is already running. though in this case the nested flow may keep running although the outer stage is already closed... so I'm not sure if this is a good direction or not...

seglo replied:
I'll attempt to answer these, but we should wait for confirmation from @jrudolph before proceeding.

> same attribute for both stages?

Yes

> default should be the 'old behaviour' hence, if downstream cancels before future/prefix completion the stage cancels, otherwise the inner flow takes charge. right?

Good question. Johannes and I didn't discuss this use case specifically, but in an effort to preserve the old behaviour I think we should do this.

> another possibility is 'going all the way' by always closing upstream upon downstream cancellation, even if the nested flow is already running. though in this case the nested flow may keep running although the outer stage is already closed... so I'm not sure if this is a good direction or not...

I was thinking about this as well. I think it makes sense to support eagerly cancelling the outer stage regardless of the running state of the inner flow either as a replacement to the new behaviour, or in addition to. I think arguments could be made for all use cases. To support this behaviour in addition to the others we could add a strategy argument to the attribute to pick a behaviour.  This would give us 3 possible behaviours (or "strategies"):

1. Cancel outer stage if inner flow is not materialized yet (old behaviour)
2. Always propagate cancellation signal through inner flow
3. Cancel outer stage (and inner flow) despite run state of inner flow

eyalfa replied:
> I'll attempt to answer these, but we should wait for confirmation from @jrudolph before proceeding.
> 
> > same attribute for both stages?
> 
> Yes
> 
> > default should be the 'old behaviour' hence, if downstream cancels before future/prefix completion the stage cancels, otherwise the inner flow takes charge. right?
> 
> Good question. Johannes and I didn't discuss this use case specifically, but in an effort to preserve the old behaviour I think we should do this.
> 
> > another possibility is 'going all the way' by always closing upstream upon downstream cancellation, even if the nested flow is already running. though in this case the nested flow may keep running although the outer stage is already closed... so I'm not sure if this is a good direction or not...
> 
> I was thinking about this as well. I think it makes sense to support eagerly cancelling the outer stage regardless of the running state of the inner flow either as a replacement to the new behaviour, or in addition to. I think arguments could be made for all use cases. To support this behaviour in addition to the others we could add a strategy argument to the attribute to pick a behaviour. This would give us 3 possible behaviours (or "strategies"):
> 
> 1. Cancel outer stage if inner flow is not materialized yet (old behaviour)
> 2. Always propagate cancellation signal through inner flow
> 3. Cancel outer stage (and inner flow) despite run state of inner flow
3rd option complicates things, what about the materialized value? it has to be somehow propagated, tut the outer stage will be closed by that time... supporting this may result with a really awkward signature and contract.
I think it'd also be very difficult to explain and document, in addition most users will use option 1, very few will use 2nd option, not sure who's going to use the 3rd option.

seglo replied:
> I think it'd also be very difficult to explain and document, in addition most users will use option 1, very few will use 2nd option, not sure who's going to use the 3rd option.

That's true.  I thought that's what you meant by this comment, but maybe I misunderstood:

> another possibility is 'going all the way' by always closing upstream upon downstream cancellation, even if the nested flow is already running. though in this case the nested flow may keep running although the outer stage is already closed... so I'm not sure if this is a good direction or not...

eyalfa commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,170 @@
+
+      object Initializing extends InHandler with OutHandler {
+        override def onPush(): Unit = {
+          throw new IllegalStateException("unexpected push during initialization")
```
@seglo , I think this is the only `Initializing` method that can be reduced to a single liner.
the rest are assigning member variables which I think shouldn't be inlined this way.

seglo replied:
:+1:

eyalfa committed:
Merge branch 'master' into akka-28016__futureFlow

eyalfa committed:
akka-28016__futureFlow: introduce a new attribute controlling handling of 'early' cancellation. modify futureFlow and its spec accordingly.

eyalfa committed:
akka-28016__futureFlow: add a test for the attribute's default value.

eyalfa committed:
akka-28016__futureFlow: FlatMapPrefix, add support for delayed/immediate downstream cancellation handling.

eyalfa commented:
@jrudolph , @seglo ,
I've implemented the attribute-controlled downstream cancellation behaviour as discussed above.
I think the implementation and tests are pretty solid, there will be required another round of updating the scaladoc and perhaps renaming the attribute. I'd appreciate your feedback.

eyalfa commented:
> @jrudolph , @seglo ,
> I've implemented the attribute-controlled downstream cancellation behaviour as discussed above.
> I think the implementation and tests are pretty solid, there will be required another round of updating the scaladoc and perhaps renaming the attribute. I'd appreciate your feedback.

@jrudolph , @seglo ,
guys?

seglo reviewed (commented):
Apologies for the delay.

eyalfa commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -0,0 +1,145 @@
+        }
+        setHandlers(in, out, new InHandler with OutHandler {
+          override def onPull(): Unit = subSink.pull()
+          override def onDownstreamFinish(cause: Throwable): Unit = subSink.cancel(cause)
```
1. this is the old behaviour (I think :sunglasses:)
2. once materialized the nested flow takes charge of the stage's ports so it makes sense it should react to downstream cancellation.

the way I see it, the attribute controls what happens before nested flow's materialization, once materialized both modes converge into one.

eyalfa replied:
if we don't propagate cancellation the nested flow will never stop which as far as I understand things will prevent the interpreter from stopping (since it'd still have running graph stages).
aside from this, this may change the nested flow behaviour and materialized value, i.e. a flow that uses `alsoToMat` in order to aggregate and report metrics/whatever will never complete its matVal if it doesn't receive a downstream cancellation signal.

jrudolph replied:
Yep, think it's fine as it is now.

eyalfa committed:
Merge branch 'master' into akka-28016__futureFlow

eyalfa committed:
akka-28016__futureFlow: compilation fixes + scalafmt after merge.

eyalfa committed:
akka-28016__futureFlow: initial docs for the new attribute + behaviour.

eyalfa commented:
@jrudolph , @seglo ,
I've started documenting the new attribute and affected operator's behaviour.
please review this docs so once approved I can 'replicate' them to the java api and the documentation pages.

seglo commented on akka-stream/src/main/scala/akka/stream/Attributes.scala:
```diff
@@ -439,11 +439,46 @@ object Attributes {
 
-  case class NestedMaterializationCancellationPolicy(delay: Boolean) extends MandatoryAttribute
+  /**
+   * Nested materialization cancellation strategy provides a way to configure the cancellation behavior of stages that materialize a nested flow at some stage.
```
```suggestion
   * Nested materialization cancellation strategy provides a way to configure the cancellation behavior of stages that materialize a nested flow.
```

eyalfa replied:
@seglo, the whole point here is that these stages may delay the nested flow's materialization - otherwise we wouldn't have to introduce this attribute.
don't u think the docs should some how mention this?

seglo replied:
Yes. I thought the trailing part of the sentence was redundant. You don't think it reads the same?

eyalfa replied:
@seglo , the attributes are separate from the actual operators they affect, hence a user reading this documentation may be confused.
I think this doc should provide the user with as much context as possible so she'd be able to understand the attribute's role.

seglo commented on akka-stream/src/main/scala/akka/stream/Attributes.scala:
```diff
@@ -439,11 +439,46 @@ object Attributes {
+   * Nested materialization cancellation strategy provides a way to configure the cancellation behavior of stages that materialize a nested flow at some stage.
+   *
+   * When cancelled before materializing their nested flows, these stages can either immediately cancel (default behaviour) without materializing the nested flow
+   * or wait for the nested materialization and then propagate the cancellation signal.
```
```suggestion
   * or wait for the nested flow to materialize and then propagate the cancellation signal through it.
```

seglo reviewed (commented):
SGTM. A few suggestions.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -90,7 +92,10 @@ import akka.util.OptionVal
+          case OptionVal.None if propagateToNestedMaterialization => downstreamCause = OptionVal.Some(cause)
+          case OptionVal.None =>
+            matPromise.failure(new NeverMaterializedException(cause))
+            super.onDownstreamFinish(cause)
```
```suggestion
            cancelStage(cause)
```

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala:
```diff
@@ -90,7 +92,10 @@ import akka.util.OptionVal
       override def onDownstreamFinish(cause: Throwable): Unit = {
         subSink match {
-          case OptionVal.None    => downstreamCause = OptionVal.Some(cause)
+          case OptionVal.None if propagateToNestedMaterialization => downstreamCause = OptionVal.Some(cause)
```
:+1: Nice, good that it was such a simple change.

jrudolph commented on akka-stream/src/main/scala/akka/stream/Attributes.scala:
```diff
@@ -439,6 +439,46 @@ object Attributes {
+   * , otherwise these stages will immediately cancel without materializing the nested flow.
+   */
+  @ApiMayChange
+  case class NestedMaterializationCancellationPolicy(propagateToNestedMaterialization: Boolean)
```
API-wise, maybe we should make the constructor private[akka] here to hide the internals? And then have constants in `NestedMaterializationCancellationPolicy`.

 * `NestedMaterializationCancellationPolicy.EagerCancellation`
 * `NestedMaterializationCancellationPolicy.PropagateToNested`

WDYT?

jrudolph replied:
Same for Java API.

eyalfa replied:
so Java API will look like: `NestedMaterializationCancellationPolicyPropagateToNested` ?
not vey welcoming, is it?

eyalfa replied:
@jrudolph can u please reply to this?

jrudolph replied:
I don't have a better suggestion, let's better be explicit.

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFutureFlowSpec.scala:
```diff
@@ -0,0 +1,527 @@
+          .withAttributes(attributes)
+          .run()
+
+        if (delayDownstreanCancellation) {
```
It's a bit unusual to structure tests like this but maybe it's ok for now.

eyalfa replied:
this was the best way I could think of to stay DRY as possible.
please notice most tests were **not** affected by the newly introduced attribute, nevertheless these still must be tested in both modes to catch any possible regression.

jrudolph replied:
Yep, I agree.

jrudolph reviewed (commented):
A few cosmetic comments but overall it looks now pretty complete. I didn't go through all the test cases again, though.

eyalfa committed:
Update akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

Co-authored-by: Johannes Rudolph <johannes.rudolph@gmail.com>

eyalfa committed:
Update akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala

Co-authored-by: Johannes Rudolph <johannes.rudolph@gmail.com>

eyalfa committed:
Update akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

Co-authored-by: Johannes Rudolph <johannes.rudolph@gmail.com>

eyalfa committed:
Update akka-stream/src/main/scala/akka/stream/Attributes.scala

Co-authored-by: Sean Glover <sean@seanglover.com>

eyalfa committed:
akka-28016__futureFlow: refactor according to @jrudolph's review.

eyalfa committed:
Update akka-stream/src/main/scala/akka/stream/Attributes.scala

Co-authored-by: Sean Glover <sean@seanglover.com>

eyalfa commented:
@jrudolph , I've refactored the 'connect' method according to your request. I'd also like your feedback about my question regarding your suggestion to about the attributes.
@jrudolph , @seglo , please let me know if you're happy with the docs so I'll copy/paste them to the java api and documentation files.

seglo commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1945,7 +1967,9 @@ trait FlowOps[+Out, +Mat] {
    * '''Cancels when''' the materialized flow cancels.
-   *  Notice that when downstream cancels prior to prefix completion, the cancellation cause is stashed until prefix completion (or upstream completion) and then handed to the materialized flow.
+   *  When downstream cancels before materialization of the nested flow, the operator's default behaviour is to cancel immediately,
+   *  this behaviour can be controlled by setting the [[akka.stream.Attributes.NestedMaterializationCancellationPolicy]] attribute to the flow.
```
```suggestion
   *  this behaviour can be controlled by setting the [[akka.stream.Attributes.NestedMaterializationCancellationPolicy]] attribute on the flow.
```

seglo commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1945,7 +1967,9 @@ trait FlowOps[+Out, +Mat] {
-   *  Notice that when downstream cancels prior to prefix completion, the cancellation cause is stashed until prefix completion (or upstream completion) and then handed to the materialized flow.
+   *  When downstream cancels before materialization of the nested flow, the operator's default behaviour is to cancel immediately,
+   *  this behaviour can be controlled by setting the [[akka.stream.Attributes.NestedMaterializationCancellationPolicy]] attribute to the flow.
+   *  When this attribute is configured to true, downstream cancellation is delayed until nested flow's materialization which is then immediately cancelled (with the original cancellation cause).
```
```suggestion
   *  When this attribute is configured to true, downstream cancellation is delayed until the nested flow's materialization which is then immediately cancelled (with the original cancellation cause).
```

seglo reviewed (commented):
I think the scaladocs look good. I just made some grammar suggestions.

I think it may make sense to copy or update the scaladocs in the Future-based operators that rely on `flatMapPrefix`, `futureFlow`, and thus this attribute: `lazyFlow` and `lazyFutureFlow`. We may also consider updating `lazyInit` and `lazyInitAsync` even though they're deprecated.

eyalfa committed:
Update akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

Co-authored-by: Sean Glover <sean@seanglover.com>

eyalfa committed:
Update akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

Co-authored-by: Sean Glover <sean@seanglover.com>

eyalfa committed:
Merge branch 'master' into akka-28016__futureFlow

eyalfa committed:
Merge remote-tracking branch 'gheyalfa/akka-28016__futureFlow' into akka-28016__futureFlow

eyalfa committed:
akka-28016__futureFlow: address mima issues.

eyalfa committed:
akka-28016__futureFlow: update scaladsl.Flow's scala doc

eyalfa committed:
akka-28016__futureFlow: update scaladsl docs.

eyalfa commented:
@seglo , @jrudolph ,
finally found the time to invest in pushing this a bit closer to completion 😎 
any way:
1. fixed mima issue (travis build does not produce a clear error in these cases, fortunately enough `mimaReportBinaryIssues` gives a clear error + remedy)
2. updated scala-dsl's scaladocs and docs
3. re. java-api: I wanted to update the docs and then it seemed to me there's some more pending work there b/c unlike the scala-dsl which is (now) based on `futureFlow` the java-dsl is based on `lazyCompletionStageFlow` which is implemented in terms of scala-dsl's `lazyFutureFlow`. I think it should be the other way around, or even better: operations should map to their scala-dsl counterpart. do.you guys agree with me?

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala:
```diff
@@ -81,63 +83,61 @@ import scala.util.{ Failure, Success, Try }
+              innerMatValue.failure(new NeverMaterializedException(ex))
+              failStage(ex)
+          }
+          setHandlers(in, out, new InHandler with OutHandler {
```
It's a bit weird that this is done even if the materialization has failed. Maybe replace the `Try {}` with a regular `try`/`catch` block.

jrudolph reviewed (commented):
LGTM, from my perspective only the two cosmetic things about the attribute API and the try/catch block are missing.

eyalfa commented:
> @seglo , @jrudolph ,
> finally found the time to invest in pushing this a bit closer to completion 😎
> any way:
> 
> 1. fixed mima issue (travis build does not produce a clear error in these cases, fortunately enough `mimaReportBinaryIssues` gives a clear error + remedy)
> 2. updated scala-dsl's scaladocs and docs
> 3. re. java-api: I wanted to update the docs and then it seemed to me there's some more pending work there b/c unlike the scala-dsl which is (now) based on `futureFlow` the java-dsl is based on `lazyCompletionStageFlow` which is implemented in terms of scala-dsl's `lazyFutureFlow`. I think it should be the other way around, or even better: operations should map to their scala-dsl counterpart. do.you guys agree with me?

@jrudolph can u address the third bullet please?

eyalfa committed:
akka-28016__futureFlow: address comment by @jrudolph.

eyalfa committed:
akka-28016__futureFlow: refactor the attribute according to review by @jrudolph.

eyalfa committed:
akka-28016__futureFlow: update attributes related docs.

eyalfa commented:
@jrudolph , @seglo , can u see what failed the build? for some reason I can't reach jnkins

eyalfa committed:
akka-28016__futureFlow: refactor attributes, refactor javadsl's futureFlow.

raboof commented:
> can u see what failed the build

Unfortunately we're having some infrastructure issues, so we can't either ;). Hoping to have them fixed soon...

jrudolph commented:
>     3\. re. java-api: I wanted to update the docs and then it seemed to me there's some more pending work there b/c unlike the scala-dsl which is (now) based on `futureFlow` the java-dsl is based on `lazyCompletionStageFlow` which is implemented in terms of scala-dsl's `lazyFutureFlow`. I think it should be the other way around, or even better: operations should map to their scala-dsl counterpart. do.you guys agree with me?

I agree, Java operations should directly map to their Scala counterpart. Maybe we can fix that in a follow up PR?

jrudolph reviewed (approved):
LGTM, how about pushing the remaining javadsl issue to a follow up PR?

jrudolph commented:
> how about pushing the remaining javadsl issue to a follow up PR?

Ah, or did you already do that now? Looks like it, great!

seglo reviewed (approved):
Great work @eyalfa.  Thanks for your patience!

eyalfa commented:
r we good to merge? 😎

jrudolph commented:
:rocket:

Crazy amount of work, @eyalfa, thanks for addressing all of our concerns!
